### DISCUSSION TITLE
OpenNARS for Minecraft world.

### DISCUSSION THREAD

**Discussion by 0x-Inf** [AGI-course/agi-course.github.io#6] [Ideas]
**Title**: Minecraft agent based on Non Axiomatic Reasoning
**Type**: Software 

### Overview 
   
The project involves using the OpenNARS system as the core logic system behind a Minecraft agent. The agent should be 
able to create 'recipes' of actions to take so as to build target structures in the minecraft world. The agent has to synthesize these 
action streams from the information it's receiving as video, audio and text. 

Minecraft is an environment with a lot of unknown information on what can be built or done, given that the agent will start off without much (probably zero) knowledge about the world, this makes it a suitable test-bed since it has the AIKR features. Particularly, the survival mode.

#### Capabilities 
 - Knowledge Representation: Convert the signals received into representations suitable for the logic engine.
 - Inference Rules: The agent has to learn in real-time how to navigate in the Minecraft world.
 - Memory: Building target structures input by an 'external' agent (a user) using the knowledge that the agent has acquired at different time-steps.

**Comment by patham9** [comment]
This is quite ambitious. Hence I recommend starting simple, for instance letting the system follow a player or NPC using events extracted from the agent's view. In simplest case by extracting some kind of (what,where) tuples from the agent's view of the environment, and then feed them to NARS. Here is a related "follow the deer" example working with ONA, which would already be great if you could get it to work in Minecraft context within the short time period of this project:
```
//go to the left when the deer is on the left side
<(<deer --> [left]> &/ ^left) =/> G>.
//go to the right when the deer is on the right side
<(<deer --> [right]> &/ ^right) =/> G>.

//the deer is on the left side
<deer --> [left]>. :|:
//get the goal achieved!
G! :|:
//expected: ^left executed
```
In more complex cases (you don't have to go that far of course), "deer left/right" information would not come from the simulation of the agent's view, but from a visual picture thereof, essentially to extract deers with an object detection model (easiest with some ConvNet such as YOLOv4).

--------------------------------------------------------------------

For inspiration, I also want to mention that NARS can categorize instances based on observable properties the detected object instances do share with the categories (abductive inference / perceptual reasoning). The following is an example of following and picking up a duck based on a visual feature of an observed instance. This kind of "abductive reasoning by comparison" is beyond what FOPL-based reasoning systems do offer. Also, the example contains causal representations with variables, which allows for follow-behavior which is not restricted to specific object categories. To date, Reinforcement Learners do not learn such abstract representations, while NARS is quite good at that. Since this example is a bit longer, I put it into a gist file: https://gist.github.com/patham9/bd8b0d8c0cfec959810747ce0f482f01

**Reply by 0x-Inf** [reply]
Thanks for the comments, they were quite helpful in narrowing down the scope of the project considering the time constraint.
